File: frontend/src/api/deviceAgent.ts
import api, { getDeviceEntryPath } from '@/api'

export type DeviceAgentRequesterKind = 'device' | 'user' | 'system'
export type DeviceAgentRequestType = 'ask' | 'diagnose' | 'delegate' | 'repair'

export interface DeviceAgentConfig {
  enabled: boolean
  display_name: string
  device_role: string
  local_context: string
  responsibilities: string
  default_provider: string
  default_model: string
  updated_at?: number | null
}

export interface DeviceAgentManifest extends DeviceAgentConfig {
  device_id: string
  hostname: string
  configured_provider?: string
  configured_model?: string
  status: string
  ai_provider?: {
    provider: string
    model: string
    available: boolean
    configured: boolean
    kind?: string
    label?: string
    error?: string | null
  }
  agent: Record<string, any>
}

export interface DeviceAgentRequester {
  kind: DeviceAgentRequesterKind
  id: string
  display_name: string
}

export interface DeviceAgentTurn {
  id: string
  session_id: string
  role: string
  requester: DeviceAgentRequester
  request_type: DeviceAgentRequestType
  instruction: string
  context: Record<string, any>
  status: string
  stage: string
  stage_label: string
  queue_task_id?: string | null
  heartbeat_at?: number | null
  result_report: Record<string, any>
  error_message?: string | null
  created_at: number
  started_at?: number | null
  finished_at?: number | null
  updated_at: number
}

export interface DeviceAgentSession {
  id: string
  local_device_id: string
  peer_device_id: string
  peer_name: string
  requester_kind: DeviceAgentRequesterKind
  title: string
  status: string
  last_turn_id?: string | null
  created_at: number
  updated_at: number
  turns?: DeviceAgentTurn[]
}

export interface DeviceAgentSessionListResponse {
  items: DeviceAgentSession[]
}

export interface DeviceAgentSessionPayload {
  requester: DeviceAgentRequester
  request_type: DeviceAgentRequestType
  instruction: string
  context?: Record<string, any>
  title?: string
}

export const fetchDeviceAgentConfig = async (entryId: string): Promise<DeviceAgentConfig> => {
  const response = await api.get(getDeviceEntryPath(entryId, '/agent/config'))
  return response.data
}

export const saveDeviceAgentConfig = async (
  entryId: string,
  payload: Partial<DeviceAgentConfig>
): Promise<DeviceAgentConfig> => {
  const response = await api.put(getDeviceEntryPath(entryId, '/agent/config'), payload)
  return response.data
}

export const fetchDeviceAgentManifest = async (entryId: string): Promise<DeviceAgentManifest> => {
  const response = await api.get(getDeviceEntryPath(entryId, '/agent/manifest'))
  return response.data
}

export const fetchDeviceAgentSessions = async (
  entryId: string,
  limit = 10
): Promise<DeviceAgentSessionListResponse> => {
  const response = await api.get(getDeviceEntryPath(entryId, '/agent/sessions'), { params: { limit } })
  return response.data
}

export const createDeviceAgentSession = async (
  entryId: string,
  payload: DeviceAgentSessionPayload
): Promise<DeviceAgentSession> => {
  const response = await api.post(getDeviceEntryPath(entryId, '/agent/sessions'), payload)
  return response.data
}

export const appendDeviceAgentTurn = async (
  entryId: string,
  sessionId: string,
  payload: DeviceAgentSessionPayload
): Promise<DeviceAgentTurn> => {
  const response = await api.post(getDeviceEntryPath(entryId, `/agent/sessions/${encodeURIComponent(sessionId)}/turns`), payload)
  return response.data
}

export const fetchDeviceAgentSession = async (
  entryId: string,
  sessionId: string
): Promise<DeviceAgentSession> => {
  const response = await api.get(getDeviceEntryPath(entryId, `/agent/sessions/${encodeURIComponent(sessionId)}`))
  return response.data
}

export const fetchDeviceAgentTurn = async (
  entryId: string,
  turnId: string
): Promise<DeviceAgentTurn> => {
  const response = await api.get(getDeviceEntryPath(entryId, `/agent/turns/${encodeURIComponent(turnId)}`))
  return response.data
}

File: backend/api/device_agent.py
from __future__ import annotations

from typing import Any, Literal, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlmodel import Session

from backend.core.access.auth import verify_api_token
from backend.core.device_agent import (
    DeviceAgentError,
    append_device_agent_turn,
    create_device_agent_session,
    get_device_agent_config,
    get_device_agent_manifest,
    get_device_agent_session,
    get_device_agent_turn,
    list_device_agent_sessions,
    save_device_agent_config,
)
from backend.db import get_session


router = APIRouter(dependencies=[Depends(verify_api_token)])


class DeviceAgentConfigRequest(BaseModel):
    enabled: Optional[bool] = None
    display_name: Optional[str] = None
    device_role: Optional[str] = None
    local_context: Optional[str] = None
    responsibilities: Optional[str] = None
    default_provider: Optional[str] = None
    default_model: Optional[str] = None


class DeviceAgentRequester(BaseModel):
    kind: Literal["device", "user", "system"] = "device"
    id: str = ""
    display_name: str = ""


class DeviceAgentTurnRequest(BaseModel):
    requester: DeviceAgentRequester = Field(default_factory=DeviceAgentRequester)
    request_type: Literal["ask", "diagnose", "delegate", "repair"] = "ask"
    instruction: str
    context: dict[str, Any] = Field(default_factory=dict)


class DeviceAgentSessionCreateRequest(DeviceAgentTurnRequest):
    title: Optional[str] = None


@router.get("/config")
def get_config(session: Session = Depends(get_session)):
    return get_device_agent_config(session)


@router.put("/config")
def put_config(payload: DeviceAgentConfigRequest, session: Session = Depends(get_session)):
    return save_device_agent_config(session, payload.model_dump(exclude_unset=True))


@router.get("/manifest")
def get_manifest(session: Session = Depends(get_session)):
    return get_device_agent_manifest(session)


@router.get("/sessions")
def list_sessions(
    limit: int = Query(30, ge=1, le=100),
    session: Session = Depends(get_session),
):
    return {"items": list_device_agent_sessions(session, limit=limit)}


@router.post("/sessions")
def create_session(payload: DeviceAgentSessionCreateRequest, session: Session = Depends(get_session)):
    try:
        return create_device_agent_session(
            session,
            requester=payload.requester.model_dump(),
            request_type=payload.request_type,
            instruction=payload.instruction,
            context=payload.context,
            title=payload.title,
        )
    except DeviceAgentError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.get("/sessions/{session_id}")
def get_session_detail(session_id: str, session: Session = Depends(get_session)):
    try:
        return get_device_agent_session(session, session_id)
    except DeviceAgentError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc


@router.post("/sessions/{session_id}/turns")
def append_turn(session_id: str, payload: DeviceAgentTurnRequest, session: Session = Depends(get_session)):
    try:
        return append_device_agent_turn(
            session,
            session_id,
            requester=payload.requester.model_dump(),
            request_type=payload.request_type,
            instruction=payload.instruction,
            context=payload.context,
        )
    except DeviceAgentError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.get("/turns/{turn_id}")
def get_turn(turn_id: str, session: Session = Depends(get_session)):
    try:
        return get_device_agent_turn(session, turn_id)
    except DeviceAgentError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
